pkg/storage/utils/eosfs/eosfs.go

```diff
@@ -487,25 +488,83 @@ func (fs *eosfs) UnsetArbitraryMetadata(ctx context.Context, ref *provider.Refer
 	return nil
 }
 
+func (fs *eosfs) setACLs(ctx context.Context, c eosclient.EOSClient, uid, gid, path string, acls []*acl.Entry) error {
+	finfo, err := c.GetFileInfoByPath(ctx, uid, gid, path)
+	if err != nil {
+		return err
+	}
+
+	a := acl.ACLs{Entries: acls}
+	aclsSerialized := a.Serialize()
+
+	attr := &eosclient.Attribute{
+		Key: "acl",
+		Val: aclsSerialized,
+	}
+	if finfo.IsDir {
+		attr.Type = SystemAttr
+	} else {
+		attr.Type = UserAttr
+	}
+
+	// system acls can only be modified by root users
+	if attr.Type == SystemAttr {
+		rootUID, rootGID, err := fs.getRootUIDAndGID(ctx)
+		if err != nil {
+			return err
+		}
+
+		uid, gid = rootUID, rootGID
+	}
+
+	if err = c.SetAttr(ctx, uid, gid, attr, true, path); err != nil {
+		return err
+	}
+	return nil
+}
+
 func (fs *eosfs) AddGrant(ctx context.Context, ref *provider.Reference, g *provider.Grant) error {
 	u, err := getUser(ctx)
 	if err != nil {
 		return errors.Wrap(err, "eosfs: no user in ctx")
 	}
 
+	uid, gid, err := fs.getUserUIDAndGID(ctx, u)
+	if err != nil {
+		return err
+	}
+
 	p, err := fs.resolve(ctx, u, ref)
 	if err != nil {
 		return errors.Wrap(err, "eosfs: error resolving reference")
 	}
 
 	fn := fs.wrap(ctx, p)
 
-	eosACL, err := fs.getEosACL(ctx, g)
-	if err != nil {
-		return err
+	// workaround until EOS has native support
+	// for persisting order of the acl entries
+	if grants.IsDenial(g) {
+		gs, err := fs.ListGrants(ctx, ref)
+		if err != nil {
+			return err
+		}
+
+		// update list of grants
+		grants.AddGrant(&gs, g)
+
+		acls, err := fs.getEosACLs(ctx, gs)
+		if err != nil {
+			return err
+		}
+
+		if err = fs.setACLs(ctx, fs.c, uid, gid, fn, acls); err != nil {
```

Comment:
I think we should continue to use setACLs function also for non denial acl, since for now when you add a new acl with the acl add command, the acls are shuffled. We risk that the denials, that we decided to keep at the end, are randomly put before a non denial acl.